File: Data_Structure/Linear/LinkedList/C++/LinkedList.cpp
#include <bits/stdc++.h>
using namespace std;
template <class T>
class LinkedList
{
private:
    struct Node
    {
        T val;
        Node *next;
    };
    Node *head;
    Node *tail;
    int count;

public:
    LinkedList()
    {
        head = nullptr;
        tail = nullptr;
        count = 0;
    }
    void push_back(T val)
    {
        Node *newVal = new Node();
        newVal->val = val;
        newVal->next = nullptr;
        if (tail == nullptr)
        {
            head = newVal;
            tail = newVal;
        }
        else
        {
            tail->next = newVal;
            tail = newVal;
        }
        count++;
    }
    void push_front(T val)
    {
        Node *newVal = new Node();
        newVal->val = val;
        newVal->next = nullptr;
        if (head == nullptr)
        {
            head = newVal;
            tail = newVal;
        }
        else
        {
            newVal->next = head;
            head = newVal;
        }
        count++;
    }
    int size()
    {
        return count;
    }
    bool isEmpty()
    {
        return count == 0;
    }
    bool contains(T val)
    {
        Node *start = head;
        while (start != nullptr)
        {
            if (start->val == val)
            {
                return true;
            }
            start = start->next;
        }
        return false;
    }
    void sort(bool ascending = true)
    {
        Node *first = head;
        Node *second = head;
        T temp;
        while (first != nullptr)
        {
            second = first->next;
            while (second != nullptr)
            {
                if (ascending)
                {
                    if (first->val > second->val)
                    {
                        temp = first->val;
                        first->val = second->val;
                        second->val = temp;
                    }
                }
                else
                {
                    if (first->val < second->val)
                    {
                        temp = first->val;
                        first->val = second->val;
                        second->val = temp;
                    }
                }
                second = second->next;
            }
            first = first->next;
        }
    }
    void pop_back()
    {
        if (count == 0)
        {
            return;
        }
        if (count == 1)
        {
            head = tail = nullptr;
            count = 0;
            return;
        }
        else
        {
            Node *start = head;
            while (start->next != tail)
            {
                start = start->next;
            }
            tail = start;
            Node *del = tail->next;
            tail->next = nullptr;
            delete del;
        }
        count--;
    }
    void pop_front(){
        if(count == 0){
            return;
        }
        if(count == 1){
            head = tail = nullptr;
            count = 0;
            return;
        }
        Node *del = head;
        head=head->next;
        del->next = nullptr;
        delete del;
        count--;
    }
    void print()
    {
        Node *start = head;
        cout << "[";
        while (start != nullptr)
        {
            cout << start->val << " ";
            start = start->next;
        }
        cout << "]\n";
    }
    void insertAt(int index,T val){
        if(index < 0 || index >= count){
            return;
        }
        if(index == 0){
            push_front(val);
            return;
        }
        if(index == count){
            push_back(val);
            return;
        }
        Node* start = head;
        index--;
        while (index--)
        {
            start = start->next;
        }
        Node *newVal = new Node;
        newVal->val = val;
        newVal->next = start->next;
        start->next = newVal;
        count++;
    }
    void deleteAt(int index){
        if(index < 0 || index >= count){
            return;
        }
        if(index == 0){
            pop_front();
            return;
        }
        if(index == count-1){
            pop_back();
            return;
        }
        index--;
        Node* start = head;
        while(index--){
            start = start->next;
        }
        Node *del = start->next;
        start->next = del->next;
        delete del;
        count--;
    }
    void clear(){
        Node *start = head;
        Node *temp;
        while (start != nullptr)
        {
            temp = start;
            start = start->next;
            delete temp;
        }
        count = 0;
        head = tail = nullptr;
    }
    ~LinkedList()
    {
        clear();
    }
};
int main()
{
    LinkedList<int> list;

    list.push_back(10);
    list.push_back(20);
    list.push_back(30);
    cout << "After push_back: ";
    list.print();

    list.push_front(5);
    list.push_front(1);
    cout << "After push_front: ";
    list.print();

    cout << "Size: " << list.size() << endl;

    cout << "Is empty: " << (list.isEmpty() ? "true" : "false") << endl;

    cout << "Contains 10: " << (list.contains(10) ? "true" : "false") << endl;
    cout << "Contains 15: " << (list.contains(15) ? "true" : "false") << endl;

    list.sort(true);
    cout << "After ascending sort: ";
    list.print();

    list.sort(false);
    cout << "After descending sort: ";
    list.print();

    list.pop_back();
    cout << "After pop_back: ";
    list.print();

    list.pop_front();
    cout << "After pop_front: ";
    list.print();

    cout << "Inserting 25 at index 2: ";
    list.insertAt(2, 25);
    list.print();

    cout << "Inserting 50 at index 0: ";
    list.insertAt(0, 50);
    list.print();

    cout << "Inserting 5 at index " << list.size() << " (end of list): ";
    list.insertAt(list.size(), 5);
    list.print();

    cout << "Deleting element at index 2: ";
    list.deleteAt(2);
    list.print();

    cout << "Deleting element at index 0: ";
    list.deleteAt(0);
    list.print();

    cout << "Deleting last element: ";
    list.deleteAt(list.size() - 1);
    list.print();

    list.clear();
    cout << "After clear, isEmpty: " << (list.isEmpty() ? "true" : "false") << endl;

    return 0;
}

File: Data_Structure/Linear/LinkedList/Java/LinkedList.java
public class LinkedList<T extends Comparable<T>> {
    private Node<T> head;
    private Node<T> tail;
    private int count;

    public LinkedList() {
        head = null;
        tail = null;
        count = 0;
    }

    public void pushBack(T val) {
        Node<T> newVal = new Node<T>();
        newVal.setValue(val);
        if (tail == null) {
            tail = newVal;
            head = newVal;
        } else {
            tail.next = newVal;
            tail = newVal;
        }
        count++;
    }

    public void pushFront(T val) {
        Node<T> newVal = new Node<T>();
        newVal.setValue(val);
        if (head == null) {
            tail = newVal;
            head = newVal;
        } else {
            newVal.next = head;
            head = newVal;
        }
        count++;
    }

    public boolean isEmpty() {
        return count == 0;
    }

    public boolean contains(T val) {
        Node<T> start = head;
        while (start != null) {
            if (start.getValue().equals(val)) {
                return true;
            }
            start = start.next;
        }
        return false;
    }

    public void sortAsc() {
        Node<T> first = head;
        Node<T> second = head;
        T temp;
        while (first != null) {
            second = first.next;
            while (second != null) {
                if (first.getValue().compareTo(second.getValue()) > 0) {
                    temp = first.getValue();
                    first.setValue(second.getValue());
                    second.setValue(temp);
                }
                second = second.next;
            }
            first = first.next;
        }
        first = null;
        second = null;
    }

    public void sortDesc() {
        Node<T> first = head;
        Node<T> second = head;
        T temp;
        while (first != null) {
            second = first.next;
            while (second != null) {
                if (first.getValue().compareTo(second.getValue()) < 0) {
                    temp = first.getValue();
                    first.setValue(second.getValue());
                    second.setValue(temp);
                }
                second = second.next;
            }
            first = first.next;
        }
        first = null;
        second = null;
    }

    public void popBack() {
        if (count == 0) {
            return;
        }
        if (count == 1) {
            head = null;
            tail = null;
            count = 0;
            return;
        }
        Node<T> start = head;
        while (start.next != tail) {
            start = start.next;
        }
        tail = start;
        tail.next = null;
        count--;
    }

    public void popFront() {
        if (count == 0) {
            return;
        }
        if (count == 1) {
            head = null;
            tail = null;
            count = 0;
            return;
        }
        Node<T> node = head;
        head = head.next;
        node.next = null;
        count--;
    }

    public void print() {
        Node<T> start = head;
        System.out.print("[");
        while (start != null) {
            System.out.print(start.getValue() + " ");
            start = start.next;
        }
        System.out.println("]");
    }

    public int size() {
        return count;
    }

    public void insertAt(int index, T val) {
        if (index < 0 || index >= count) {
            return;
        }
        if (index == 0) {
            pushFront(val);
            return;
        }
        if (index == count) {
            pushBack(val);
            return;
        }
        Node<T> start = head;
        index--;
        while (index > 0) {
            start = start.next;
            index--;
        }
        Node<T> newVal = new Node<T>();
        newVal.setValue(val);
        newVal.next = start.next;
        start.next = newVal;
        count++;
    }

    public void deleteAt(int index) {
        if (index < 0 || index >= count) {
            return;
        }
        if (index == 0) {
            popFront();
            return;
        }
        if (index == count) {
            popBack();
            return;
        }
        Node<T> start = head;
        index--;
        while (index > 0) {
            start = start.next;
            index--;
        }
        Node<T> del = start.next;
        start.next = del.next;
        del = null;
        count--;
    }

    public void reverse() {
        Node<T> curr = head;
        Node<T> next = curr.next;
        Node<T> prev = null;
        Node<T> newTail = head;
        while (next != null) {
            curr.next = prev;
            prev = curr;
            curr = next;
            next = next.next;
        }
        curr.next = prev;
        head = curr;
        tail = newTail;
    }
    public void clear() {
        // Thanks to Garbage Collector 
        head = null;
        tail = null;
        count = 0;
    }
}
